analyzers/tests/SonarAnalyzer.UnitTest/TestCases/MemberInitializedToDefault.CSharp11.cs

```diff
@@ -1,32 +1,75 @@
-    public IntPtr field6 = 0x0000000000000000;      // Noncompliant
-    public UIntPtr field7 = UIntPtr.Zero;           // Noncompliant
-    public IntPtr field8 = new IntPtr(0);           // Noncompliant
+    public IntPtr field5 = IntPtr.Zero; // Noncompliant
+    public IntPtr field6 = 0x0000000000000000; // Noncompliant
+    public UIntPtr field7 = UIntPtr.Zero; // Noncompliant
+    public IntPtr field8 = new IntPtr(0); // Noncompliant
     public IntPtr field9 = new IntPtr(staticField); // Compliant
 
-    public UIntPtr field10 = new UIntPtr(0);        // Noncompliant
-    public UIntPtr field11 = new();                 // Noncompliant
-    public UIntPtr field12 = new UIntPtr { };       // Noncompliant
+    public UIntPtr field10 = new UIntPtr(0); // Noncompliant
+    public UIntPtr field11 = new();  // Noncompliant
+    public UIntPtr field12 = new UIntPtr { }; // Noncompliant
 
-    public IntPtr Property1 { get; set; } = 0;      // Noncompliant
-    public IntPtr Property2 { get; set; } = 42;     // Compliant
+    public IntPtr Property1 { get; set; } = 0; // Noncompliant
+    public IntPtr Property2 { get; set; } = 42; // Compliant
 
-    public UIntPtr Property3 { get; init; } = 0;    // Noncompliant
-    public UIntPtr Property4 { get; init; } = 42;   // Compliant
+    public UIntPtr Property3 { get; init; } = 0; // Noncompliant
+    public UIntPtr Property4 { get; init; } = 42; // Compliant
 
     public IntPtr Property5 { get; set; } = 0 * 20 - 0; // FN - Expression is not evaluated
     public UIntPtr Property6 { get; set; } = 0 * 20 - 0; // FN - Expression is not evaluated
 }
+
+public readonly struct FooStruct
+{
+    public FooStruct() { }
+
+    public int Value { get; init; } // Compliant
+    public bool BoolValue { get; init; } // Compliant
+
+    public int ValueDefault { get; init; } = 0; // Noncompliant
+    public bool BoolValueDefault { get; init; } = false; // Noncompliant
+}
+
+public struct BarStruct
+{
+    public int someField = 0; // Noncompliant Initializing this field is optional for C# 11 due to the auto-default-struct C# feature https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-11#auto-default-struct
+    public BarStruct(int dummy) { }
+}
+
+public struct FooBarStruct
+{
+    public int someField; // Compliant - does not raise a CS0171 issue anymore due to the auto-default-struct C# 11 feature https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-11#auto-default-struct
+    public required int someRequiredField; // Compliant - does not raise a CS0171 and the compiler will default this to 0.
+
+    public FooBarStruct(int dummy) { }
+}
+
+public class ClassWithRequiredProperties
+{
+    public required int RequiredProperty { get; init; }
+    public int AnotherProperty { get; set; }
+
+    public ClassWithRequiredProperties()
+    {
+        RequiredProperty = 0; // FN - the compiler will set this automatically to 0
+    }
```

Comment:
There is also another missing piece in the rule which relates to `required`. Before `required` was introduced you faced this problem if you wanted to use an object initializer:
```cs
#nullable enable
// You want to use object initialization and nullable
var c = new C() { MyProperty = WhatEver() }; 

public class C
{
// warning CS8618: Non-nullable property 'MyProperty' must contain a non-null value when exiting constructor. Consider declaring the property as nullable.
    public object MyProperty { get; init; }
    //            ^^^^^^^^^^
}
// So you changed it to this:
public class C
{
    public object MyProperty { get; init; } = default!;
// To silence the warning you use a field initializer and the suppression operator. We should not warn in the case ! is present.
}
```
With `required` this workaround is no longer needed. We need an issue for the rule about the `!` operator for `default` and `null`. This is related to C# 8.

Reply:
🤦 I messed up this example.

This rule raises issues for the field, event field, and property declaration.

Regarding the required properties - the compiler complains to the caller  - not in the class declaration.
So we should still raise in-class - but not when the caller sets the property from the constructor.

Example:
```csharp
using System.Diagnostics.CodeAnalysis;

var test = new ClassWithRequiredProperties(); 
test.RequiredProperty; // It's automatically set to zero, SetsRequiredMembers does all the job.

public class ClassWithRequiredProperties
{
    public required int RequiredProperty { get; init; }
    public int AnotherProperty { get; set; }

    [SetsRequiredMembers]
    public ClassWithRequiredProperties()
    {
    }
}
```

but 

```csharp
using System.Diagnostics.CodeAnalysis;

var test = new ClassWithRequiredProperties(); // Raises CS9035 error
var test2 = new ClassWithRequiredProperties() { RequiredProperty = 0 };  // we should not raise here

public class ClassWithRequiredProperties
{
    public required int RequiredProperty { get; init; } = 0; // even by setting this still there's a CS9035 so we should raise here.
    public int AnotherProperty { get; set; }

}
```
I'll add these examples.

Reply:
> There is also another missing piece in the rule which relates to the required. Before required was introduced you faced this problem if you wanted to use an object initializer:

I'll add this test to the next PR since it's not strictly related to C# 11.